### ISSUE TITLE
Function Objects do not seem to be supported for Hotkeys (Crashes)

### ISSUE THREAD

**Issue by evilC** [amazing-andrew/AutoHotkey.Interop#9]
```
            ahk.ExecRaw(@"
            Test(){
                fn := Func(""Test2"")
                hotkey, F12, % fn
            }

            Test2(){
                msgbox Test2
            }
            ");

            ahk.ExecRaw(@"Test()");
```
= `VsHost32.exe has stopped working`

**Comment by evilC** [comment]
Ignore, seems to be fine if you replace the packaged AutoHotkey.dll with the latest one

**Comment by evilC** [comment]
It appears that the problem is to do with `SendPipeMessage`  
```
            ahk.ExecRaw(@"
                F11::
                    Tooltip, % ""HERE - "" A_TickCount
                    return
                
                F12::
                    SendPipeMessage(""HERE - "" A_TickCount)
                    return
            ");
```
The F11 hotkey displays a tooltip reliably on every press, but the F12 hotkey does not.

**Comment by evilC** [comment]
The problem appears to be due to the fact that SendPipeMessage is a two-way communication:

```
SendPipeMessage(strMessage) {
	global A__PIPECLIENT
	A__PIPECLIENT.write(strMessage)
	sleep, 100
	A__PIPECLIENT_RESULT := A__PIPECLIENT.read()
	sleep, 100
	return A__PIPECLIENT_RESULT
}
```

Those sleep 100s clearly are causing the problem.  

If I comment out everything after `A__PIPECLIENT.write(strMessage)` and ignore the return value, I get callbacks for all hotkeys.  

All I actually want is one-way communication from the script to the host C# code (ie an Action instead of a Func) I tried to change the code in this way, but had no joy.  

I have been chatting to HotkeyIt about this, and he reckons this could also be done with ObjShare, although he hasn't done C# before and I have not got a clue how this all works. We'll keep beavering away, but if you are about then I could really do with some help on this one.

**Comment by amazing-andrew** [comment]
How does the program behave, if you just take the sleeps out ?

**Comment by evilC** [comment]
Yeah it seems OK if I take them out and don't try to do anything with a return value, but I don't know enough about it to know if it is safe to do so, or could be more performant without it.  
HotkeyIt also thought it would be better to implement communication via ObjShare, but neither of use are really sure how to implement this in C# - any ideas?  
https://github.com/HotKeyIt/ahkdll/blob/master/source/resources/reslib/ObjShare.ahk

**Comment by amazing-andrew** [comment]
The CSharp code acts as a Server by hosting the Named Pipe, used for communication.
The AHK code acts as a client by making the connection only when it needs to send something.

Currently the named pipe server behaves like the following:

- .NET : It opens the named pipe and waits for a connection to be made.
- AHK : When you use SendPipeMessage in AHK, it makes a connection the pipe message, ahk sends the message as a string. AHK then waits in a loop for a response back from the named pipe.
- .NET : the named pipe server receives the string, and then sends the string to the named pipe handler function. the function runs and returns a string. It sends the response string by writing the string response back into the named pipe. Only after a response has been sent back to the AHK client does the server waiting for a new connection. This does mean that the named pipe server does not handle more than 1 connection at a time. 
- AHK : once it receives the response back it returns it as a result to the SendPipeMessage function.

Overall the .NET and AHK processes are waiting on each other before they continue, so they block each other's threads. The sleeps are probably put in by me for testing purposes. Sometimes I find AHK is more stable when you give it some time around important pieces of code.

If you find that your system handles the calls without the two sleeps well, I don't see an issue with removing them. I agree it may perform better. 

Also I've looked at ObjShare, and please correct me if i'm wrong. But it looks like a way to share an AHK object between AHK Threads. I'm not sure how this can translate sending a message to the hosting environment and returning a result back from it. Do you have an example in any programming language of the ObjShare being used to call a function in the hosting environment, outside of AHK?  I can look at this and perhaps find a way to replicate it in CSharp.

**Comment by evilC** [comment]
Not got time to read this all right now, will get back to it, but for now here is an example of inter-thread communication using a function object wrapped with ObjShare:  

Main thread launching second thread, passing callback func object wrapped with ObjShare:  
https://github.com/evilC/UCR/blob/master/Classes/Profile.ahk#L106  

Child thread getting function object:  
https://github.com/evilC/UCR/blob/master/Threads/ProfileInputThread.ahk#L14  

Child thread using function object:  
https://github.com/evilC/UCR/blob/master/Threads/ProfileInputThread.ahk#L195

**Comment by jaryn-kubik** [comment]
So I've been toying around with this and the SendPipeMessage really is kinda unreliable.
However from what I found out, since the ahkdll runs in the same process, you don't even need any sort of IPC for communicating, you can just directly call a C# delegate with DllCall.

Example:
```csharp
void AHKCallback(string s)
{
   Console.WriteLine(s);
}

[UnmanagedFunctionPointer(CallingConvention.StdCall)]
delegate void AHKDelegate([MarshalAs(UnmanagedType.LPWStr)]string s);

AHKDelegate ahkDelegate = AHKCallback;

void writeCrapInConsoleOnPressingA()
{
   IntPtr ptr = Marshal.GetFunctionPointerForDelegate(ahkDelegate);
   var ahk = AutoHotkeyEngine.Instance;
   ahk.ExecRaw(@"a::
      DllCall(" + ptr + @", ""Str"", ""whatever you wanna pass to AHKCallback"")
      return");
}
```

**Comment by amazing-andrew** [comment]
I like this method a lot. Should be simple and less hassles implementing this type communication between the two worlds.

**Comment by evilC** [comment]
So if we switched to this method, does that mean that all the pipes code could be removed?  
I need to have lots of copies of AHK running - up to a minimum of about 11.  
Why so many? Because I need one for "Bind Mode" (Has hotkeys to all keyboard and mouse buttons declared) which is used to detect which key the end-user wishes to use, then one copy of AHK per user profile.  
This way, it is easy to turn on/off sets of hotkeys - you just Suspend/Unsuspend the copies of AHK as appropriate.  
I managed to alter the code to not be a singleton, and in order to do so had to comment out a bunch of stuff to do with pipes.  
Am hoping maybe switching to this method will also make this possible without having to have my own custom fork.
